go/arrow/ipc/writer.go

```diff
@@ -724,6 +796,34 @@ func (w *recordEncoder) getZeroBasedValueOffsets(arr arrow.Array) (*memory.Buffe
 	return voffsets, nil
 }
 
+func (w *recordEncoder) rebaseDenseUnionValueOffsets(arr *array.DenseUnion, offsets, lengths []int32) *memory.Buffer {
+	// this case sucks. Because the offsets are different for each
+	// child array, when we have a sliced array, we need to re-base
+	// the value offsets for each array! ew.
+	unshiftedOffsets := arr.RawValueOffsets()
+	codes := arr.RawTypeCodes()
+
+	shiftedOffsetsBuf := memory.NewResizableBuffer(w.mem)
+	shiftedOffsetsBuf.Resize(arrow.Int32Traits.BytesRequired(arr.Len()))
+	shiftedOffsets := arrow.Int32Traits.CastFromBytes(shiftedOffsetsBuf.Bytes())
+
+	// offsets may not be ascending, so we need to find out the start offset for each child
```

Comment:
The [spec](https://arrow.apache.org/docs/format/Columnar.html#dense-union) says:
> The respective offsets for each child value array must be in order / increasing.

Reply:
I see you basically copied the C++ implementation. I'll open a JIRA to simplify it.

Reply:
I filed ARROW-17339.

Reply:
Yea i thought that was strange too. Given filing that card, i'll simplify this to take advantage of assuming the offsets are monotonically increasing.

Reply:
simplified logic so now we can do only 1 loop through and when we find the -1 for the offset we have found the initial offset for that child and use that as our zero offset.